[JIT] hook back on container magic methods on save and load

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32796 [JIT] hook back on container magic methods on save and load**

In order to continue to support exposing magic methods of modules which inherit from nn containers we copy over the python magic methods to their scripted versions. This PR extends that behavior to saved & loaded modules by saving the fully qualified name of a module when it is scripted. When we load a scripted module, we look up the python methods to copy on the original module it was saved from.

A follow up PR to this could fix the problem we had of isinstance(MyMod, torch.jit.script(MyMod)) returning false now that we save the qualified name of module on scripting.

Commit:
hook back on container magic methods on save and load

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 2d07dc83:


* **4/4** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=2d07dc83a3f5e73b191d8fd58820ca05557d368a).

### :detective: 4 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4374819) pytorch_linux_xenial_py2_7_9_test (1/4)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92868201) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92868201))
<details>
<summary>
<code>Jan 29 23:46:31 RuntimeError: test_quantization failed!</code>
</summary>

```
Jan 29 23:46:31 Ran 36 tests in 50.904s 
Jan 29 23:46:31  
Jan 29 23:46:31 FAILED (errors=16, skipped=2) 
Jan 29 23:46:31  
Jan 29 23:46:31 Generating XML reports... 
Jan 29 23:46:31 Traceback (most recent call last): 
Jan 29 23:46:31   File "test/run_test.py", line 457, in <module> 
Jan 29 23:46:31     main() 
Jan 29 23:46:31   File "test/run_test.py", line 450, in main 
Jan 29 23:46:31     raise RuntimeError(message) 
Jan 29 23:46:31 RuntimeError: test_quantization failed! 
Jan 29 23:46:31 + cleanup 
Jan 29 23:46:31 + retcode=1 
Jan 29 23:46:31 + set +x 
Jan 29 23:46:31 =================== sccache compilation log =================== 
Jan 29 23:46:31 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 29 23:46:31 Compile requests                 10 
Jan 29 23:46:31 Compile requests executed         8 
Jan 29 23:46:31 Cache hits                        2 
Jan 29 23:46:31 Cache misses                      6 
Jan 29 23:46:31 Cache timeouts                    0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4375026) pytorch_linux_backward_compatibility_check_test (2/4)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92869539) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92869539))
<details>
<summary>
<code>Jan 29 23:47:34 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.</code>
</summary>

```
Jan 29 23:47:34 processing existing schema:  aten::values(Dict(str, t) self) -> (t[](*)) 
Jan 29 23:47:34 processing existing schema:  aten::values(Dict(int, t) self) -> (t[](*)) 
Jan 29 23:47:34 processing existing schema:  aten::values(Dict(float, t) self) -> (t[](*)) 
Jan 29 23:47:34 processing existing schema:  aten::values(Dict(Tensor, t) self) -> (t[](*)) 
Jan 29 23:47:34 processing existing schema:  aten::view(Tensor(a) self, int[] size) -> (Tensor(a)) 
Jan 29 23:47:34 processing existing schema:  aten::where(Tensor condition) -> (Tensor[]) 
Jan 29 23:47:34 processing existing schema:  aten::where.self(Tensor condition, Tensor self, Tensor other) -> (Tensor) 
Jan 29 23:47:34 processing existing schema:  aten::zeros_like(Tensor self, *, int? memory_format=None) -> (Tensor) 
Jan 29 23:47:34 processing existing schema:  aten::zeros_like.dtype(Tensor self, *, int dtype, int layout, Device device, bool pin_memory=False, int? memory_format=None) -> (Tensor) 
Jan 29 23:47:34 processing existing schema:  prim::Constant() -> (...) 
Jan 29 23:47:34 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.  
Jan 29 23:47:34  
Jan 29 23:47:34 Broken ops: [ 
Jan 29 23:47:34 	quantized::linear_relu_dynamic_fp16(Tensor X, Tensor W_prepack) -> (Tensor Y) 
Jan 29 23:47:34 ] 
Jan 29 23:47:34 + cleanup 
Jan 29 23:47:34 + retcode=1 
Jan 29 23:47:34 + set +x 
Jan 29 23:47:34 =================== sccache compilation log =================== 
Jan 29 23:47:34 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 29 23:47:34 Compile requests                 0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4375035) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_simple_test (3/4)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92870357) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92870357))
<details>
<summary>
<code>Jan 29 23:48:17 RuntimeError: test_jit_simple failed!</code>
</summary>

```
Jan 29 23:48:17 Ran 2369 tests in 72.525s 
Jan 29 23:48:17  
Jan 29 23:48:17 FAILED (failures=1, errors=5, skipped=79, expected failures=1) 
Jan 29 23:48:17  
Jan 29 23:48:17 Generating XML reports... 
Jan 29 23:48:17 Traceback (most recent call last): 
Jan 29 23:48:17   File "test/run_test.py", line 457, in <module> 
Jan 29 23:48:17     main() 
Jan 29 23:48:17   File "test/run_test.py", line 450, in main 
Jan 29 23:48:17     raise RuntimeError(message) 
Jan 29 23:48:17 RuntimeError: test_jit_simple failed! 
Jan 29 23:48:17 + cleanup 
Jan 29 23:48:17 + retcode=1 
Jan 29 23:48:17 + set +x 
Jan 29 23:48:17 =================== sccache compilation log =================== 
Jan 29 23:48:17 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 29 23:48:17 Compile requests                 2 
Jan 29 23:48:17 Compile requests executed        0 
Jan 29 23:48:17 Cache hits                       0 
Jan 29 23:48:17 Cache misses                     0 
Jan 29 23:48:17 Cache timeouts                   0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4375036) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_legacy_test (4/4)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92871452) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92871452))
<details>
<summary>
<code>Jan 29 23:48:59 RuntimeError: test_jit_legacy failed!</code>
</summary>

```
Jan 29 23:48:59 Ran 2369 tests in 127.684s 
Jan 29 23:48:59  
Jan 29 23:48:59 FAILED (failures=1, errors=5, skipped=63, expected failures=1) 
Jan 29 23:48:59  
Jan 29 23:48:59 Generating XML reports... 
Jan 29 23:48:59 Traceback (most recent call last): 
Jan 29 23:48:59   File "test/run_test.py", line 457, in <module> 
Jan 29 23:48:59     main() 
Jan 29 23:48:59   File "test/run_test.py", line 450, in main 
Jan 29 23:48:59     raise RuntimeError(message) 
Jan 29 23:48:59 RuntimeError: test_jit_legacy failed! 
Jan 29 23:48:59 + cleanup 
Jan 29 23:48:59 + retcode=1 
Jan 29 23:48:59 + set +x 
Jan 29 23:48:59 =================== sccache compilation log =================== 
Jan 29 23:49:00 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 29 23:49:00 Compile requests               250 
Jan 29 23:49:00 Compile requests executed        0 
Jan 29 23:49:00 Cache hits                       0 
Jan 29 23:49:00 Cache misses                     0 
Jan 29 23:49:00 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "[JIT] hook back on container magic methods on save and load"


In order to continue to support exposing magic methods of modules which inherit from nn containers we copy over the python magic methods to their scripted versions. This PR extends that behavior to saved & loaded modules by saving the fully qualified name of a module when it is scripted. When we load a scripted module, we look up the python methods to copy on the original module it was saved from.

A follow up PR to this could fix the problem we had of isinstance(MyMod, torch.jit.script(MyMod)) returning false now that we save the qualified name of module on scripting.


[ghstack-poisoned]

Commit:
Update on "[JIT] hook back on container magic methods on save and load"


In order to continue to support exposing magic methods of modules which inherit from nn containers we copy over the python magic methods to their scripted versions. This PR extends that behavior to saved & loaded modules by saving the fully qualified name of a module when it is scripted. When we load a scripted module, we look up the python methods to copy on the original module it was saved from.

A follow up PR to this could fix the problem we had of isinstance(MyMod, torch.jit.script(MyMod)) returning false now that we save the qualified name of module on scripting.


[ghstack-poisoned]

Commit:
Update on "[JIT] hook back on container magic methods on save and load"


In order to continue to support exposing magic methods of modules which inherit from nn containers we copy over the python magic methods to their scripted versions. This PR extends that behavior to saved & loaded modules by saving the fully qualified name of a module when it is scripted. When we load a scripted module, we look up the python methods to copy on the original module it was saved from.

A follow up PR to this could fix the problem we had of isinstance(MyMod, torch.jit.script(MyMod)) returning false now that we save the qualified name of module on scripting.


[ghstack-poisoned]

eellison commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -259,13 +259,11 @@ struct SourceImporterImpl : public Resolver,
     std::vector<Def> methods;
     std::vector<ResolverPtr> resolvers;
     std::vector<Assign> attributes;
     std::vector<Assign> constants;
```
maybe `unmangled_qual_name` is better ?

eellison reviewed (commented):
i also tested this by saving a Module to a file, and loading it in a different file where the module was not defined and it still worked

suo reviewed (changes requested):
We currently have a distinction between the sugared Python properties of a Module (represented by `ConcreteModuleType`) and unsugared, first-class properties of a Module (represented by `ClassType`).

Our rule to date has been to serialize only first-class properties, and leave the sugaredness behind. This is desirable for a few reasons:
1. It means loaded modules behave the same in Python and C++
2. It means that the serialized module fully describes the module's behavior; i.e. we have no implicit dependency on external code for loading a module.
3. It simplifies module semantics, i.e. we can think about a python ScriptModule as a wrapper over a C++ module that does all real stuff.

I think all three are important; 2 is the scariest to me—e.g. in the current PR if we jitter the qualified name of, say, nn.Sequential (either through a change in how we compute qualified names, or by just moving around the class), we break BC.

In general, I think we need to avoid things that try preserve to a Python-y part of the module's state on save and "re-link" it back to a (potentially very different) Python environment on load.

I'm not sure what is the best solution here, maybe we should brainstorm together about it.

eellison commented:
Previously we had discussed in a meeting supporting `isinstance(MyMod, torch.jit.script(MyMod))` by keeping around the qualified name and looking up the module on load. It's also what pickle does.

This PR extends that logic to this specific use case. This will also let us do better error messages. If `torch.nn.containers.Sequential` name is no longer resolvable than we'll have broken BC for reasons other than this PR.
